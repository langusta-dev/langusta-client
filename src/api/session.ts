import { rest } from '~/helpers/api'

import type { AuthPayload } from '~/types/session'

export const logIn = async (data: AuthPayload) => {
  const response = await rest.post<{ token: string }>('/auth/login', data)
  return { ...response, data: response.data?.token || null }
}

export const register = async (data: AuthPayload) => {
  const response = await rest.post<{ token: string }>('/auth/register', data)
  return { ...response, data: response.data?.token || null }
}
