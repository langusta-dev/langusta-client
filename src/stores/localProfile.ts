import { acceptHMRUpdate, defineStore } from 'pinia';

import { getJwt } from '~/composables/jwt';

export const useLocalProfileStore = defineStore('localProfile', () => {
  let isLocalProfileEnabled = $(
    useLocalStorage<boolean>('enable-local-profile', false)
  );

  const enableLocalProfile = () => {
    isLocalProfileEnabled = true;
  };

  const disableLocalProfile = () => {
    isLocalProfileEnabled = false;
  };

  whenever(getJwt, disableLocalProfile);

  return {
    isLocalProfileEnabled: computed(() => isLocalProfileEnabled),
    enableLocalProfile,
    disableLocalProfile,
  };
});

if (import.meta.hot) {
  import.meta.hot.accept(
    acceptHMRUpdate(useLocalProfileStore, import.meta.hot)
  );
}
