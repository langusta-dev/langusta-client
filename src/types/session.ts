export interface LogInPayload {
  username: string;
  password: string;
}

export interface RegisterPayload {
  email: string;
  username: string;
  firstname: string;
  lastname: string;
  password: string;
}
