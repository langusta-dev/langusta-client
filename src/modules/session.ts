import { useSession } from '~/stores/session'

import { jwt } from '~/composables/jwt'

import type { InstallModule } from '~/types/modules'

export const install: InstallModule = ({ isClient, router }) => {
  if (!isClient) {
    return
  }

  router.beforeEach((to) => {
    if (to.meta.auth) {
      const session = useSession()

      if (!session.isAuth) {
        return '/login'
      }
    }

    return true
  })

  watch(jwt, (v) => {
    if (!v) {
      router.push('/login')
    }
  })
}
