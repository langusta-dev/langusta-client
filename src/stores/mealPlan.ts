import { acceptHMRUpdate, defineStore } from 'pinia';

import { useSynchronizableArray } from '~/composables/dataSync';

import {
  fetchUserMealPlans,
  uploadMealPlans,
  deleteMealPlansByIds,
} from '~/api/mealPlan';

import { useLocalProfileStore } from './localProfile';

export const useMealPlanStore = defineStore('recipeCollection', () => {
  const localProfileStore = useLocalProfileStore();

  const mealPlanInitializer = () =>
    localProfileStore.isLocalProfileEnabled ? [] : fetchUserMealPlans();

  const {
    isInSync: areMealPlansInSync,
    syncPromise: mealPlansSyncPromise,
    state: mealPlans,
    getById: getMealPlanById,
    push: addMealPlan,
  } = useSynchronizableArray(
    'mealPlans',
    mealPlanInitializer,
    uploadMealPlans,
    deleteMealPlansByIds
  );

  return {
    areMealPlansInSync,
    mealPlansSyncPromise,
    mealPlans,
    getMealPlanById,
    addMealPlan,
  };
});

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useMealPlanStore, import.meta.hot));
}
