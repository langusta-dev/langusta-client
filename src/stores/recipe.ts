import { acceptHMRUpdate, defineStore } from 'pinia';
import { v4 as uuid } from 'uuid';

import { isOnline } from '~/composables/online';

import { fetchRecipes, uploadRecipes } from '~/api/recipe';

import { useLocalProfileStore } from './localProfile';

import type { Recipe, RecipeId, EditableRecipe } from '~/types/recipe';

export const useRecipeStore = defineStore('recipes', () => {
  const localProfileStore = useLocalProfileStore();

  const { state: fetchedRecipes } = $(
    useAsyncState<Recipe[]>(fetchRecipes, [])
  );

  const localRecipes = $(useLocalStorage<Recipe[]>('recipes', []));

  const recipes = $computed(() => [...localRecipes, ...fetchedRecipes]);

  const recipesPerId = $computed(
    () => new Map(recipes.map((recipe) => [recipe.id, recipe]))
  );

  const getRecipeById = (id: RecipeId) => recipesPerId.get(id) || null;

  const recipesToUpload = $(useLocalStorage<Recipe[]>('recipes-to-upload', []));

  const addRecipe = (recipe: EditableRecipe) => {
    const now = new Date().toString();

    const newRecipe = { id: uuid(), createdAt: now, updatedAt: now, ...recipe };

    localRecipes.push(newRecipe);

    if (!localProfileStore.isLocalProfileEnabled) {
      recipesToUpload.push(newRecipe);
    }
  };

  watch([$$(recipesToUpload), isOnline], async () => {
    if (isOnline.value && recipesToUpload.length) {
      await uploadRecipes(recipesToUpload);
      recipesToUpload.splice(0, recipesToUpload.length);
    }
  });

  // TODO
  // clear `localRecipes` and `recipesToUpload` on user change

  return { recipes: computed(() => recipes), getRecipeById, addRecipe };
});

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useRecipeStore, import.meta.hot));
}
