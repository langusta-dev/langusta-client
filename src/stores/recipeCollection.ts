import { acceptHMRUpdate, defineStore } from 'pinia';

import { useSynchronizableArray } from '~/composables/dataSync';

import {
  fetchUserRecipeCollections,
  uploadRecipeCollections,
  deleteRecipeCollectionsByIds,
} from '~/api/recipeCollection';

import { useLocalProfileStore } from './localProfile';

export const useRecipeCollectionStore = defineStore('recipeCollection', () => {
  const localProfileStore = useLocalProfileStore();

  const collectionInitializer = () =>
    localProfileStore.isLocalProfileEnabled ? [] : fetchUserRecipeCollections();

  const {
    isInSync: areCollectionsInSync,
    state: collections,
    ownedState: ownedCollections,
    getById: getCollectionById,
    isOwnedById: isCollectionOwnedById,
    push: addCollection,
    editById: editCollectionById,
    deleteById: deleteCollectionById,
  } = useSynchronizableArray(
    'recipe-collections',
    collectionInitializer,
    uploadRecipeCollections,
    deleteRecipeCollectionsByIds
  );

  return {
    areCollectionsInSync,
    collections,
    ownedCollections,
    getCollectionById,
    addCollection,
    isCollectionOwnedById,
    editCollectionById,
    deleteCollectionById,
  };
});

if (import.meta.hot) {
  import.meta.hot.accept(
    acceptHMRUpdate(useRecipeCollectionStore, import.meta.hot)
  );
}
