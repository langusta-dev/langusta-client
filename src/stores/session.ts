import { acceptHMRUpdate, defineStore } from 'pinia';

import { jwt, setJwt, unsetJwt } from '~/composables/jwt';

import * as api from '~/api/session';

import { useLocalProfileStore } from './localProfile';

import type { RestResponse } from '~/types/api';
import type { JwtTokenPayload } from '~/types/jwt';
import type { LogInPayload, RegisterPayload } from '~/types/session';

export const useSessionStore = defineStore('session', () => {
  const localProfileStore = useLocalProfileStore();

  const isAuth = $computed(
    () => !!jwt.value || localProfileStore.isLocalProfileEnabled
  );

  const logOut = () => {
    unsetJwt();

    if (localProfileStore.isLocalProfileEnabled) {
      localProfileStore.disableLocalProfile();
    }
  };

  const _authenticate = async (
    tokenPayloadResponse: Promise<RestResponse<JwtTokenPayload>>
  ) => {
    if (isAuth) {
      logOut();
    }

    const { data } = await tokenPayloadResponse;

    if (data?.token) {
      setJwt(data.token);
    }
  };

  const logIn = (payload: LogInPayload) => _authenticate(api.logIn(payload));

  const register = async (payload: RegisterPayload) =>
    _authenticate(api.register(payload));

  return { isAuth: $$(isAuth), logIn, logOut, register };
});

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useSessionStore, import.meta.hot));
}
