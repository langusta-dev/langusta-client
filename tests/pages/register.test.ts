import Register from '~/pages/register.vue';

import { flushPromises, mount } from '@vue/test-utils';

import { flushDelayedPromises } from '~test-utils';

import { useSessionStore } from '~/stores/session';

describe('register page', () => {
  const email = 'some@email.abc';
  const username = 'some-username';
  const firstname = 'some-firstname';
  const lastname = 'some-lastname';
  const password = 'some-password';

  it('should render', () => {
    const wrapper = mount(Register);

    const inputs = wrapper.findAll('input');
    expect(inputs).toHaveLength(6);

    const buttons = wrapper.findAll('button');
    expect(buttons).toHaveLength(1);

    expect(wrapper.text()).toContain('register.email');
    expect(wrapper.text()).toContain('register.username');
    expect(wrapper.text()).toContain('register.firstname');
    expect(wrapper.text()).toContain('register.lastname');
    expect(wrapper.text()).toContain('register.password');
    expect(wrapper.text()).toContain('register.repeat_password');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should register', async () => {
    const wrapper = mount(Register);

    const inputs = wrapper.findAll('input');

    await inputs[0].setValue(email);
    await inputs[1].setValue(username);
    await inputs[2].setValue(firstname);
    await inputs[3].setValue(lastname);
    await inputs[4].setValue(password);
    await inputs[5].setValue(password);

    const button = wrapper.find('button');
    await button.trigger('click');

    const sessionStore = useSessionStore();

    expect(sessionStore.register).toHaveBeenCalledOnce();
    expect(sessionStore.register).toHaveBeenCalledWith({
      email,
      username,
      firstname,
      lastname,
      password,
    });
  });

  it('should show error on failed registration', async () => {
    vi.useFakeTimers();
    const wrapper = mount(Register);

    const inputs = wrapper.findAll('input');

    await inputs[0].setValue(email);
    await inputs[1].setValue(username);
    await inputs[2].setValue(firstname);
    await inputs[3].setValue(lastname);
    await inputs[4].setValue(password);
    await inputs[5].setValue(password);

    const sessionStore = useSessionStore();

    // @ts-expect-error it's readonly
    sessionStore.isAuth = false;

    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();
    await flushDelayedPromises();

    expect(sessionStore.register).toHaveBeenCalledOnce();

    // @ts-expect-error it's readonly
    sessionStore.isAuth = false;

    expect(wrapper.html()).toContain('register.registration_failed_error');
  });

  it('should show error on invalid email input', async () => {
    vi.useFakeTimers();
    const wrapper = mount(Register);

    const inputs = wrapper.findAll('input');
    const emailInput = inputs[0];

    const invalidEmail = 'not-an-email';

    await emailInput.setValue(invalidEmail);

    await emailInput.trigger('blur');
    await flushDelayedPromises();

    expect(wrapper.html()).toContain('register.invalid_email_error');
    expect(emailInput.element.value).toBe(invalidEmail);
  });

  it('should show error on too short password input', async () => {
    vi.useFakeTimers();
    const wrapper = mount(Register);

    const inputs = wrapper.findAll('input');
    const passwordInput = inputs[4];

    await passwordInput.setValue(
      'x'.repeat(import.meta.env.VITE_PASSWORD_MIN_LENGTH - 1)
    );

    await passwordInput.trigger('blur');
    await flushDelayedPromises();

    expect(wrapper.html()).toContain('register.password_too_short_error');
    expect(passwordInput.element.value).toBe('');
  });

  it('should show error on not matching passwords', async () => {
    vi.useFakeTimers();
    const wrapper = mount(Register);

    const inputs = wrapper.findAll('input');
    const passwordInput = inputs[4];
    const repeatPasswordInput = inputs[5];

    await inputs[0].setValue(email);
    await inputs[1].setValue(username);
    await inputs[2].setValue(firstname);
    await inputs[3].setValue(lastname);
    await passwordInput.setValue(password);
    await repeatPasswordInput.setValue(password + 'x');

    const sessionStore = useSessionStore();

    const button = wrapper.find('button');
    await button.trigger('click');
    await flushDelayedPromises();

    expect(sessionStore.register).not.toHaveBeenCalled();

    expect(wrapper.html()).toContain('register.passwords_not_equal_error');

    expect(passwordInput.element.value).toBe('');
    expect(repeatPasswordInput.element.value).toBe('');
  });
});
