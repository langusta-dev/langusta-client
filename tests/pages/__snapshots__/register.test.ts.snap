// Vitest Snapshot v1

exports[`register page > should render 1`] = `
"<div _flex=\\"~ col\\" _items-center=\\"\\" _h-full=\\"\\">
  <div _text=\\"center 3xl\\" _fw600=\\"\\" _p=\\"t6 b8\\">register.title</div>
  <div _flex=\\"~ col\\" _items-center=\\"\\" _text-center=\\"\\" _gap4=\\"\\">
    <transition-group-stub enterfromclass=\\"op0\\" leavetoclass=\\"op0\\" moveclass=\\"!transition-300 z-1\\" enteractiveclass=\\"!pointer-events-none !transition-300\\" leaveactiveclass=\\"absolute !pointer-events-none !transition-300\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
      <!--v-if--><input placeholder=\\"register.email\\" type=\\"text\\" class=\\"p-x2 p-y1.5 w80 max-w-full rounded placeholder:text-sm placeholder:text-primary-contrast/70 placeholder:fw500 border-1 outline outline-2 outline-offset--1 outline-transparent hover:outline-accent !focus:outline-accent-focus transition transition-property-[background,border,outline] border-primary-contrast/30 bg-primary-contrast/2\\" autocomplete=\\"false\\" _text-center=\\"\\"><input placeholder=\\"register.username\\" type=\\"text\\" class=\\"p-x2 p-y1.5 w80 max-w-full rounded placeholder:text-sm placeholder:text-primary-contrast/70 placeholder:fw500 border-1 outline outline-2 outline-offset--1 outline-transparent hover:outline-accent !focus:outline-accent-focus transition transition-property-[background,border,outline] border-primary-contrast/30 bg-primary-contrast/2\\" autocomplete=\\"false\\" _text-center=\\"\\"><input placeholder=\\"register.firstname\\" type=\\"text\\" class=\\"p-x2 p-y1.5 w80 max-w-full rounded placeholder:text-sm placeholder:text-primary-contrast/70 placeholder:fw500 border-1 outline outline-2 outline-offset--1 outline-transparent hover:outline-accent !focus:outline-accent-focus transition transition-property-[background,border,outline] border-primary-contrast/30 bg-primary-contrast/2\\" autocomplete=\\"false\\" _text-center=\\"\\"><input placeholder=\\"register.lastname\\" type=\\"text\\" class=\\"p-x2 p-y1.5 w80 max-w-full rounded placeholder:text-sm placeholder:text-primary-contrast/70 placeholder:fw500 border-1 outline outline-2 outline-offset--1 outline-transparent hover:outline-accent !focus:outline-accent-focus transition transition-property-[background,border,outline] border-primary-contrast/30 bg-primary-contrast/2\\" autocomplete=\\"false\\" _text-center=\\"\\"><input placeholder=\\"register.password\\" type=\\"password\\" class=\\"p-x2 p-y1.5 w80 max-w-full rounded placeholder:text-sm placeholder:text-primary-contrast/70 placeholder:fw500 border-1 outline outline-2 outline-offset--1 outline-transparent hover:outline-accent !focus:outline-accent-focus transition transition-property-[background,border,outline] border-primary-contrast/30 bg-primary-contrast/2\\" autocomplete=\\"false\\" _text-center=\\"\\"><input placeholder=\\"register.repeat_password\\" type=\\"password\\" class=\\"p-x2 p-y1.5 w80 max-w-full rounded placeholder:text-sm placeholder:text-primary-contrast/70 placeholder:fw500 border-1 outline outline-2 outline-offset--1 outline-transparent hover:outline-accent !focus:outline-accent-focus transition transition-property-[background,border,outline] border-primary-contrast/30 bg-primary-contrast/2\\" autocomplete=\\"false\\" _text-center=\\"\\">
      <div _flex=\\"~ col\\" _items-center=\\"\\" _text-center=\\"\\" _gap2=\\"\\"><button class=\\"bg-accent-interactive text-accent-interactive-contrast p-x8 p-y1\\" disabled=\\"\\" _rounded-full=\\"\\" _fw600=\\"\\" _text-center=\\"\\" _transition-all=\\"\\" _cursor-pointer=\\"\\" _select-none=\\"\\" _disabled:pointer-events-none=\\"\\" _disabled:op40=\\"\\" _flex-center=\\"\\">register.sign_up</button></div>
      <div>register.has_account <a href=\\"/login\\" class=\\"text-primary-contrast-interactive\\" _inline-block=\\"\\" _underline=\\"\\" _fw500=\\"\\" _transition-colors=\\"\\" _cursor-pointer=\\"\\" _select-none=\\"\\">register.log_in</a></div>
    </transition-group-stub>
  </div>
</div>"
`;
